### DISCUSSION TITLE
Getting Information For a Specific User

### DISCUSSION THREAD

**Discussion by dadams39** [microsoftgraph/msgraph-sdk-go#158] [Q&A] [answered]
There is a lot of information about obtaining information from the user level using the following pattern:
`result, err := graphClient.Me().MessagesById(&messageId).Get(options)`
However, this doesn't work for tenant-level access where there are multiple users. Assuming the application is set up correctly, the following code snippet does not conclude with an Iterable object that takes me through the user's messages. 
```
user := pageItem.(models.Userable)
result := user.GetMessages()
pageIterator, err2 := msgraphcore.NewPageIterator(result, adapter, models.CreateMessageCollectionResponseFromDiscriminatorValue)
```
I am certain that I need to send a new query through the API in order to receive this information. However, I do not know where I can obtain the pathParameters to use the [NewUserItemRequestBuilder](https://github.com/microsoftgraph/msgraph-sdk-go/blob/df8bddf3bea3870aae4dbed1cbe7f9fca2df9bdd/users/users_request_builder.go#L75).

**Comment by baywet** [comment]
I'm not sure I understand the question here. Is the question "how can I list all messages for a single user?" or "how can I list all messages for all users on the tenant?"

**Reply by dadams39** [reply]
More the second one... I really only wanted two out _n_ users on the tenant.
Answering
> "how can I list all messages for all users on the tenant?"

would be very helpful.

**Reply by baywet** [reply]
First you need to build a list of all the user ids, this can be achieved [through the users endpoint](https://docs.microsoft.com/en-us/graph/api/user-list?view=graph-rest-1.0&tabs=http) and with the help of the page iterator you already know how to use.
Tip: use a select query parameter so you retrieve only the Id, this will save time on the service, network and client application.
Once you've built a list of the ids, iterate through them and list the messages with the page iterator.
(use a select here too for the fields you need).

```golang
result, err := graphClient.UsersById(&UserId).Messages().Get(options)
// pass the result to the page iterator
```

Note: depending on the scenario you're trying to achieve, you might run into some volume/throttling issues. If you're trying to sync the data to an external system, using change notifications and change tracking will be more adapted. If you're trying to export data for analytics/DLP scenario, using Microsoft Graph Data Connect might be better. If you're running an export for a migration, using the Office 365 migration APIs might be better.

**Reply by dadams39** [reply]
I was looking in the Userable.go and User.go and I did not specifically see a get method for the Id. Does it exist in a different file?

**Reply by baywet** [reply]
It comes from the parent [entityable/entity](https://github.com/microsoftgraph/msgraph-sdk-go/blob/df8bddf3bea3870aae4dbed1cbe7f9fca2df9bdd/models/entityable.go#L11)

**Comment by dadams39** [comment, marked as answer]
From @baywet  
> First you need to build a list of all the user ids, this can be achieved [through the users endpoint](https://docs.microsoft.com/en-us/graph/api/user-list?view=graph-rest-1.0&tabs=http) and with the help of the page iterator you already know how to use.
Tip: use a select query parameter so you retrieve only the Id, this will save time on the service, network and client application.
Once you've built a list of the ids, iterate through them and list the messages with the page iterator.
(use a select here too for the fields you need).
```
result, err := graphClient.UsersById(&UserId).Messages().Get(options)
// pass the result to the page iterator
```
This is the answer. **Key Point:** All of the information is not available just because the User struct has fields such as calendars and messages. A separate query must be run in order to retrieve certain data.  Optimizing those queries for your application is left to the developer's discretion.
